paho/client.go

```diff
@@ -641,133 +655,140 @@ func (c *Client) Subscribe(ctx context.Context, s *Subscribe) (*Suback, error) {
 		}
 	}
 
-	c.debug.Printf("subscribing to %+v", s.Subscriptions)
-
-	subCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
-	defer cf()
-	cpCtx := &CPContext{subCtx, make(chan packets.ControlPacket, 1)}
+	respInternal := make(chan packets.ControlPacket)
+	resp := make(chan packets.ControlPacket, 1)
 
 	sp := s.Packet()
 
-	mid, err := c.MIDs.Request(cpCtx)
+	var err error
+	sp.PacketID, err = c.MIDs.Request(&CPContext{context.TODO(), respInternal})
 	if err != nil {
 		return nil, err
 	}
-	defer c.MIDs.Free(mid)
-	sp.PacketID = mid
 
-	c.debug.Println("sending SUBSCRIBE")
-	if _, err := sp.WriteTo(c.Conn); err != nil {
-		return nil, err
-	}
-	c.debug.Println("waiting for SUBACK")
-	var sap packets.ControlPacket
+	c.debug.Printf("subscribing to %+v", s.Subscriptions)
 
-	select {
-	case <-subCtx.Done():
-		if ctxErr := subCtx.Err(); ctxErr != nil {
-			c.debug.Println(fmt.Sprintf("terminated due to context: %v", ctxErr))
-			return nil, ctxErr
+	c.workers.Add(1)
+	go func() {
+		defer c.MIDs.Free(sp.PacketID)
+		defer c.workers.Done()
+		defer close(resp)
+
+		c.debug.Println("sending SUBSCRIBE for %d", sp.PacketID)
+		if _, err := sp.WriteTo(c.Conn); err != nil {
+			c.errors.Printf("failed to send SUBSCRIBE for %d: %w", sp.PacketID, err)
+			return
 		}
-	case sap = <-cpCtx.Return:
-	}
 
-	if sap.Type != packets.SUBACK {
-		return nil, fmt.Errorf("received %d instead of Suback", sap.Type)
-	}
-	c.debug.Println("received SUBACK")
+		var receivedPacket packets.ControlPacket
 
-	sa := SubackFromPacketSuback(sap.Content.(*packets.Suback))
-	switch {
-	case len(sa.Reasons) == 1:
-		if sa.Reasons[0] >= 0x80 {
-			var reason string
-			c.debug.Println("received an error code in Suback:", sa.Reasons[0])
-			if sa.Properties != nil {
-				reason = sa.Properties.ReasonString
-			}
-			return sa, fmt.Errorf("failed to subscribe to topic: %s", reason)
-		}
-	default:
-		for _, code := range sa.Reasons {
-			if code >= 0x80 {
-				c.debug.Println("received an error code in Suback:", code)
-				return sa, fmt.Errorf("at least one requested subscription failed")
-			}
+		select {
+		case <-c.stop:
+			c.errors.Printf("SUBSCRIBE operation for %d stopping due to client stop channel", sp.PacketID)
+			return
+		case receivedPacket = <-respInternal:
+		}
+
+		if receivedPacket.Type != packets.SUBACK {
+			reasonString := fmt.Sprintf(
+				"expected SUBACK for %d, got %s",
+				sp.PacketID,
+				receivedPacket.PacketType(),
+			)
+			c.errors.Printf("disconnecting due to server protocol error. %s", reasonString)
+			go c.Disconnect(&Disconnect{
+				ReasonCode: packets.DisconnectProtocolError,
+				Properties: &DisconnectProperties{ReasonString: reasonString},
+			})
+			return
 		}
-	}
+		resp <- receivedPacket
+	}()
 
-	return sa, nil
+	select {
+	case <-subCtx.Done():
+		c.debug.Printf("adding SUBSCRIBE for %d to OrphanedOps due to context cancellation: %w", sp.PacketID, subCtx.Err())
+		c.OrphanedOps.AddOp(resp)
+		return nil, subCtx.Err()
```

Comment:
Important detail: if the context is cancelled, the channel gets added to `OrphanedOps`, which holds onto the channels for orphaned operations and calls the user-specified callback if a result comes in. Same deal with Unsubscribe() and Publish()